draco/core/io.py

```diff
@@ -352,6 +342,328 @@ def process(self):
+            pos = np.concatenate(pos, axis=0)
+            z = np.concatenate(z)
+            obj_id = np.concatenate(obj_id)
+
+            # create common wavelength axis
+            loglam_all = np.concatenate(loglam)
+            llam_max = loglam_all.max()
+            llam_min = loglam_all.min()
+            dllam = np.min(np.abs(np.diff(loglam[0])))
+            # N.B. all spectra must share a common resolution
+            for ll in loglam:
+                this_dllam = np.min(np.abs(np.diff(ll)))
+                if np.abs(dllam - this_dllam) / dllam > 1e-3:
+                    self.log.warn("Not all spectra use same gridding.")
+            nlam = int(np.round((llam_max - llam_min) / dllam)) + 1
+            lam = np.power(10, llam_min + np.arange(nlam) * dllam)
+
+            # sort data into a single array
+            delta_common = np.zeros((len(delta), nlam), dtype=delta[0].dtype)
+            weight_common = np.zeros((len(weight), nlam), dtype=weight[0].dtype)
+            t0 = time()
+            for i in range(len(delta)):
+                ind = np.argmin(
+                    np.abs(np.power(10, loglam[i])[:, np.newaxis] - lam), axis=1
+                )
+                delta_common[i, ind] = delta[i]
+                weight_common[i, ind] = weight[i]
+            self.log.debug(f"Done sorting spectra. ({time() - t0:.1f} s)")
+            delta = delta_common
+            weight = weight_common
+
+            # weights are inverse variance estimate with a redshift-dependent bias correction
+            # (eq. 7 in du Mas des Bourboux et al)
+
+            # can't just concatenate because lam axis is different
+            num_objects = len(pos)
+            num_pix = len(lam)
+
+            self.log.debug("rank 0 done reading files")
+            self.log.debug(f"delta shape {delta.shape}")
+
+        else:
+            num_objects = None
+            num_pix = None
+            delta = None
+            weight = None
+
+        self.comm.Barrier()
+
+        # Broadcast the size of the catalog to all ranks
+        num_objects = self.comm.bcast(num_objects, root=0)
+        num_pix = self.comm.bcast(num_pix, root=0)
+        self.log.debug(f"Reading spectra for {num_objects} objects.")
+
+        # broadcast the axes
+        if self.comm.rank != 0:
+            pos = np.zeros((num_objects, 2), dtype=np.float64)
+            z = np.zeros(num_objects, dtype=np.float64)
+            lam = np.zeros(num_pix, dtype=np.float64)
+            obj_id = np.zeros(num_objects, dtype=np.int32)
+        self.comm.Bcast(pos, root=0)
+        self.comm.Bcast(z, root=0)
+        self.comm.Bcast(lam, root=0)
+        self.comm.Bcast(obj_id, root=0)
+        self.log.debug(f"(rank {self.comm.rank}) got axes.")
+
+        # convert wavelength (A) to frequency (MHz)
+        freq = units.c / lam * 1e4
+
+        # scatter the distributed datasets over frequency
+        def scatter_array(array, root=0, axis=0):
```

Comment:
I think you can easily leverage existing MPIArray routines to implement a scatter. Basically you add an extra length-1 axis at the front on rank=0, and then pretend that it is a (badly) distributed MPIArray, and then redistribute to axis=0, and then peel off the spurious first axis. Something like:
```
shape = array.shape if self.comm.rank == 0 else None
shape = self.comm.bcast(shape, root=0)
array_ext = array[np.newaxis, ...] if self.comm.rank == 0 else np.empty((0,) + shape, dtype=np.float64)
arr0 = mpiarray.MPIArray.wrap(array_ext, axis=0)
arr_dist = arr0.redistribute(axis=1)[0]
```

Reply:
This is definitely better than reimplementing a scatter directly in here.

Reply:
I see, yeah that is nicer... Well at least I got to practice some MPI calls.